### DISCUSSION TITLE
2x performance increase with new nvidia drivers

### DISCUSSION THREAD

**Discussion by ghost** [vladmandic/sdnext#1145] [General]
Found this today: https://www.reddit.com/r/StableDiffusion/comments/13q4ku4/nvidia_2x_performance_improvement_for_stable/

This reads like its tensorrt but its coming straight from Nvidia. Models will need to be converted just like with tensorrt. Release tomorrow

Apparently AMD cards support it too: https://gpuopen.com/amd-microsoft-directml-stable-diffusion/

**Comment by tkiikt** [comment]
[https://blogs.nvidia.com/blog/2023/05/23/microsoft-build-nvidia-ai-windows-rtx/](https://blogs.nvidia.com/blog/2023/05/23/microsoft-build-nvidia-ai-windows-rtx/)
> On May 24, we’ll release our latest optimizations in Release 532.03 drivers that combine with Olive-optimized models to deliver big boosts in AI performance. Using an Olive-optimized version of the Stable Diffusion text-to-image generator with the popular Automatic1111 distribution, performance is improved over 2x with the new driver.

[Optimize DirectML performance with Olive](https://devblogs.microsoft.com/directx/optimize-directml-performance-with-olive/)
[Stable Diffusion Optimization with DirectML](https://github.com/microsoft/Olive/tree/main/examples/directml/stable_diffusion)

**Comment by Tillerz** [comment]
:-o Does `stable_diffusion.py --optimize` work on windows?

**Comment by Centurion-Rome** [comment]
"The stable diffusion models are large, and the optimization process is resource intensive. It is recommended to run optimization on a system with a minimum of 16GB of memory (preferably 32GB). Expect optimization to take several minutes (especially the U-Net model)."

**Reply by ghost** [reply]
I also have only 8 but I think renting a GPU will be feasible for conversion. You could convert your favourite models for a few dollars and then be done with it.

But as with all speed booster news in SD I dont want to be too optimistic about this yet.

**Reply by Silanda** [reply]
> I also have only 8 but I think renting a GPU will be feasible for conversion. You could convert your favourite models for a few dollars and then be done with it.

That says RAM, not VRAM.

**Comment by mark-wd** [comment]
Let's see how it fares in real world usage.

**Comment by vladmandic** [comment]
Yeah, I don't have much faith in this.

"Olive" optimized means compiling model from Torch to ONNX and then using new ONNX backend with built-in tensor cores support.
Which would mean that entire SD ecosystem should stop using Torch/CUDA/ROCm/etc and switch to ONNX - highly unlikely.

And if going down that path, I'd much rather support TensorRT directly instead of ONNX unless someone says that ONNX is next-best-thing and will completely surpass Torch in the near-future.

Microsoft uses ONNX internally in Azure for ML-as-a-Service, so this makes nVidida a much more interesting choice for MS and I get why nVidida would do it (why wouldn't they?).

**Reply by Tillerz** [reply]
Ah poo, I had high hopes for a speed bump. :( I mean, with an 4090 I cannot complain, but if there's MORE, why not have it. :-p

**Reply by maikelsz** [reply]
> Yeah, I don't have much faith in this.
> 
> "Olive" optimized means compiling model from Torch to ONNX and then using new ONNX backend with built-in tensor cores support. Which would mean that entire SD ecosystem should stop using Torch/CUDA/ROCm/etc and switch to ONNX - highly unlikely.
> 
> And if going down that path, I'd much rather support TensorRT directly instead of ONNX unless someone says that ONNX is next-best-thing and will completely surpass Torch in the near-future.
> 
> Microsoft uses ONNX internally in Azure for ML-as-a-Service, so this makes nVidida a much more interesting choice for MS and I get why nVidida would do it (why wouldn't they?).

my GPUs are AMD and I work on Windows. I am not going to use Linux or spend on NVIDIA. If this gives me more performance in my configurations, fine, I'll use it, others can continue using CUDA/ROCm

**Comment by lshqqytiger** [comment]
I had a look this yesterday.
It generated a 512x512 image with 2.03it/s on my RX 5700 XT. (1.5it/s on vladmandic's)
And it will be faster if you are using NVIDIA card and on the latest driver.
But it is for ONNX Runtime, so it is difficult to apply it right now.

**Comment by aifartist** [comment]
I never got the Olive stuff to work on native Ubuntu.
I have been doing my own conversions of my models from PYT/DiffUsers -> onnx -> optimized onnx -> TRT using my own pipeline.
The speed is great 90+ it/s but if I compare side by side the quality does suffer.  AITemplate is slower than TRT but the quality seems good but I haven't done an exhaustive comparison.

I have a theory that it is the ONNX intermediate stage that is messing up the quality during the conversion.  I see many large values clamped to INT32 and subnormal FP16 value warnings.   A1111 does crisp images using fp16.  I don't see why a compiled TRT model using fp16 should be worse unless during certain intermediate processing some weights were screwed up.  I can't get confirmation on this theory although I did post an issue on the onnx github.

**Reply by Tillerz** [reply]
Is SD way faster on Linux than windows with the same quality?

**Reply by mark-wd** [reply]
> I never got the Olive stuff to work on native Ubuntu. I have been doing my own conversions of my models from PYT/DiffUsers -> onnx -> optimized onnx -> TRT using my own pipeline. The speed is great 90+ it/s but if I compare side by side the quality does suffer. AITemplate is slower than TRT but the quality seems good but I haven't done an exhaustive comparison.
> 
> I have a theory that it is the ONNX intermediate stage that is messing up the quality during the conversion. I see many large values clamped to INT32 and subnormal FP16 value warnings. A1111 does crisp images using fp16. I don't see why a compiled TRT model using fp16 should be worse unless during certain intermediate processing some weights were screwed up. I can't get confirmation on this theory although I did post an issue on the onnx github.

90 it/s is near real time rendering...

**Reply by vladmandic** [reply]
> Is SD way faster on Linux than windows with the same quality?

No. Its the same, there is a slight difference in interrupt handling as well as vram overhead due to how drivers work, but at the end no performance difference.

**Reply by aifartist** [reply]
> > I never got the Olive stuff to work on native Ubuntu. I have been doing my own conversions of my models from PYT/DiffUsers -> onnx -> optimized onnx -> TRT using my own pipeline. The speed is great 90+ it/s but if I compare side by side the quality does suffer. AITemplate is slower than TRT but the quality seems good but I haven't done an exhaustive comparison.
> > I have a theory that it is the ONNX intermediate stage that is messing up the quality during the conversion. I see many large values clamped to INT32 and subnormal FP16 value warnings. A1111 does crisp images using fp16. I don't see why a compiled TRT model using fp16 should be worse unless during certain intermediate processing some weights were screwed up. I can't get confirmation on this theory although I did post an issue on the onnx github.
> 
> 90 it/s is near real time rendering...

That is what I'm working towards.

**Comment by tkiikt** [comment]
a test run on Windws: [https://youtu.be/r-ZYOlZTLDA](https://youtu.be/r-ZYOlZTLDA)

**Reply by vladmandic** [reply]
great, but irrelevant as onnx is as far as you can get from current implementation which is 100% torch. even if you recompile the model and get onnxruntime working, what about everything else that makes this ecosystem so amazing? cross-optimizations, extensions, upscalers, loras, embeddings, etc...that cannot be ported to onnx, you might as well start with a complete new project from scratch.

**Reply by tkiikt** [reply]
I agree, if other changes are needed to Lora and extensions besides SD, it is a hassle not only for the developer but also for the user. Is it cost-effective to do so...

**Reply by Tillerz** [reply]
At least we know it works, and we can now wait and watch what's happening in the community. :)

**Reply by vladmandic** [reply]
if we're going by how not-fast-at-all the tensorrt integation has happened, it would be pretty bad.
i'm still hoping to get a clean tensorrt integration - this entire onnx story is just a round-around way to get to tensor cores anyhow.

**Reply by aifartist** [reply]
> if we're going by how not-fast-at-all the tensorrt integation has happened, it would be pretty bad. i'm still hoping to get a clean tensorrt integration - this entire onnx story is just a round-around way to get to tensor cores anyhow.

One competitor to TRT, AITemplate just checked in support for ControlNet two days ago.  I'm trying to figure out whether I should change my focus to AIT.  But I understand the TRT code path.  So should I fix the ONNX issues or bypass it, if possible, or jump ship to AIT, Olive, or whatever comes out tomorrow.  :-)

I even used ChatGPT to give me a python front end which handles HTTP txt2img POST requests which A1111's api uses and returns an image for my TRTInfer backend.

**Reply by vladmandic** [reply]
I never understood why nVidia doesn't integrate tensor cores support into Cuda so it "just works" in Torch. I mean, TensorRT is great if you want to work in C++, but in Python, its one of those why would I? And ONNX with TensorCores is one more step further from the ecosystem. Great for isolated test cases and running benchmarks, but porting entire SD ecosystem to ONNX? Not a chance.
AITemplate sounds interesting due to its native compile, but I'm yet to see a working example on anything non-trivial.

**Reply by aifartist** [reply]
> I never understood why nVidia doesn't integrate tensor cores support into Cuda so it "just works" in Torch. I mean, TensorRT is great if you want to work in C++, but in Python, its one of those why would I? And ONNX with TensorCores is one more step further from the ecosystem. Great for isolated test cases and running benchmarks, but porting entire SD ecosystem to ONNX? Not a chance. AITemplate sounds interesting due to its native compile, but I'm yet to see a working example on anything non-trivial.

I got AIT to work in VoltaML end to end from ckpt file to AIT engine.  Maybe 63 it/s(?) I forget.   I did some gens earlier today because I'm disappointed in TRT image quality which is obvious when comparing side by side.  And can be horrible on some images.

I'm not sure what you mean by non-trivial.  And they added support for control net a few days back.

The tensor core things still confuses me.  I was told if you use bp16 the work gets automatically sent to the Tensor Cores.  Also, another person told me the tensor cores are not independent cores.  That are like ALU's that can be accessed by the CUDA cores.  Perhaps similar to AVX except there isn't a 1-1 association between one cuda and one tensor unit.  There are in a shared pool.  Probably a poor explanation but it made sense when it was described to me.

**Reply by vladmandic** [reply]
> I got AIT to work in VoltaML end to end from ckpt file to AIT engine.

define end-to-end?
txt2img is easy. img2img a bit more difficult. combining with hires fix (which is both txt2img and img2img) more so. inpainting? yet another story. then come lora's, lyco's, etc. and then complex stuff like controlnet and multidiffusion. that's what i mean by eco-system.

> I was told if you use bp16 the work gets automatically sent to the Tensor Cores

well, yeah, cudo will auto-use tensors cores - if the phase of the moon aligns with stars :)
first, it has to be mixed precision and either fp16 or bf16 (not just bf16), but then it has to be a veeeery specific tensor and layer shape - so you end up not using them 99% of the time. most of "compile" options actually go through the model execution and change shapes ever so slightly so they can actually be executed using tensor cores. that's the primary reason why "compile" actually takes that long as simple change of format would be much faster.

oh, and you have no idea if tensors cores are utilized or not unless you run app through nvidia's profiler (dlprof) which is its own can of worms. there is no gpu monitoring utility (from nvidia or 3rd party that shows anything useful).

**Reply by aifartist** [reply]
> oh, and you have no idea if tensors cores are utilized or not unless you run app through nvidia's profiler (dlprof) which is its own can of worms. there is no gpu monitoring utility (from nvidia or 3rd party that shows anything useful).

I've made the same complaint to NVidia about being able to monitor Tensor core usage.  They might say that since they aren't real cores and more like execution units accessed by the real cuda cores so you need a profiler.

Your explanation now makes it clear to me just what NVidia/MSFT might have finally done well, with Tensor Core usage, but this makes me even more angry that NVidia is supporting MSFT's vendor lock-in with the DirectX requirement.  In theory, there is something on Linux that contains DirectX('wine' for one example) but it is unclear if this could be made to get Olive/DirectML to work.  NVidia shouldn't need MSFT to figure out how to access their own hardware properly.

Also, this might now explain why, when I use TRT, I once saw 96 it/s but currently I'm only getting 70+.  I probably had the sweet spot of various package versions, settings, the wind at my back, and the alignment of the celestial bodies.

**Reply by vladmandic** [reply]
when it comes to gaming, consumer market dictates the terms  

when it comes to ai, cloud deployments dictate the terms as that is where large majority of sales happen  
and largest consumers of such deployed HW are not using torch or anything like that - they are using ML-as-a-Service and don't care about the backend  

The fact that Azure ML internally uses ONNX? Who cares  
And Google Cloud TPU uses heavily modified TensorFlow for their own HW  

ML as I-as-a-Service comes second  
then corporate deployments  
and then in the faaar distance there is consumer market  

Torch has decent consumer support primarily because of large research community

**Reply by aifartist** [reply]
The two guys I've assisted the most are just above the consumer market but hardly large corporate entities.  One of them has 12 servers with 24 4090's in his garage and is looking for a building to build a server farm to expand on the increasing demands of serving images.  Paid subs, good growth, and a very sound business guy who knows what he is doing.  He calls me regularly to brainstorm on features and to work to improve his performance.

If I understand thing correctly my 70 it/s, about 2X faster, is likely just Tensor RT, but when I've hit 96 it might be because everything was just perfect to bring the Tensor Cores into use.

**Reply by vladmandic** [reply]
i think even to hit 70it/s cores were used for some ops, i don't think tensorrt is that much faster than torch on its own (it is faster, but not 50%). but then stars aligned and you get tensor cores getting fully utilized and hit 96 which is a near-magic number for 4090.

i'm really curios how does aitemplate perform?

**Reply by aifartist** [reply]
> i think even to hit 70it/s cores were used for some ops, i don't think tensorrt is that much faster than torch on its own (it is faster, but not 50%). but then stars aligned and you get tensor cores getting fully utilized and hit 96 which is a near-magic number for 4090.
> 
> i'm really curios how does aitemplate perform?

VoltaML does 40.9 it/s on models in general.
With AIT 512x512 Euler_a 25 steps I get:
100%|████████████| 25/25 [00:00<00:00, 70.70it/s]
100%|████████████| 25/25 [00:00<00:00, 72.25it/s]
100%|████████████| 25/25 [00:00<00:00, 72.22it/s]
100%|████████████| 25/25 [00:00<00:00, 72.00it/s]
which is as good as I'm seeing right now with my messed up TRT env.

When I first used VML it was quite hacky and experimental.  They've come a long ways.
The AIT path seems sound.  If you wanted to try it I could walk you through install, setup, and use on Linux if you have Linux.

**Reply by vladmandic** [reply]
> The AIT path seems sound. If you wanted to try it I could walk you through install, setup, and use on Linux if you have Linux.

sure!

**Reply by vladmandic** [reply]
btw, one concern is purely teorhetical - model needs to be "baked in" for all those optimizations to work, which means that lora's cannot be applied on top of it. if you want to use lora/lyco, you'd need to merge it with model (and you get sort-of dreambooth) and then bake that. not really feasible in production if you want to use any kind of trained stuff.

**Reply by aifartist** [reply]
> > The AIT path seems sound. If you wanted to try it I could walk you through install, setup, and use on Linux if you have Linux.
> 
> sure!

First you need AIT
git clone --recursive https://github.com/facebookincubator/AITemplate
cd AITemplate
python3 -m venv \`pwd\`/venv
source venv/bin/activate
cd python
python3 setup.py build   # The instructions say use "bdist_wheel" but that doesn't work.
cd ../..

git clone https://github.com/voltaML/voltaML-fast-stable-diffusion
cd voltaML-fast-stable-diffusion
python3 -m venv \`pwd\`/venv
source venv/bin/activate
pip3 install ../AITemplate/python/build   # main installs everything but this.
python3 main.py

Wait till the install finishes and connect browser to http://127.0.0.1:5003/
Find tiny DOWNLOAD icon on left edge, choose model and download.
Click tiny left hand icon for "txt2img" can click the "Load Model" button.  Click "Load" on the model you want.  Note the 3 tabs "PyTorch", "AITemplate" and "Extra",  Right now you want a PyTorch model.
Return to txt2img and generate an image changing standard setting like sampler or batch count.  On a fast system you show get around 40.

AITemplate:
Click the small Accelerate icon on the left edge.
Select the model at the bottom.
Perhaps raise the number of cpu threads to do the compile from 8 to 24 depending on your cpu and how idle your system is.  This should just affect the compile time.
Click "Accelerate" and wait.
When done return to "txt2img" and click Load Model and click on the AITemplate tab.
You should see a model now there.  Load it and generate an image.

If anything goes wrong describe it to me and wait till I see your message and respond.
I'd like to see what kind of speed you see on your setup.

**Reply by lshqqytiger** [reply]
> btw, one concern is purely teorhetical - model needs to be "baked in" for all those optimizations to work, which means that lora's cannot be applied on top of it. if you want to use lora/lyco, you'd need to merge it with model (and you get sort-of dreambooth) and then bake that. not really feasible in production if you want to use any kind of trained stuff.

I agree. There will be many things which cannot be migrated to ONNX ecosystem.

**Comment by mark-wd** [comment]
That's what I was going to ask... if the underlying technology can't just be harnessed without needing to go about this whole route.

**Comment by mark-wd** [comment]
https://www.reddit.com/r/StableDiffusion/comments/13u0kha/optimization_comparison_in_a1111_13_tensorrt_vs/
